{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Speed benchmark of five aligners for different thread counts\n",
    "\n",
    "This benchmark is testing the alignment speed of 5 aligners and pseudo-aligners. Pseudo-aligners are using k-mers (shorter nucleotide fragments) of length 31 in this case. Through an expectation maximization the k-mers from a read are used to predict the most likely transcript from which they originate. This notebook is ment to be executed from the docker image maayanlab/alignmentbenchmark. All dependencies should be installed.\n",
    "\n",
    "### Pseudo-aligners\n",
    "- kallisto\n",
    "- Salmon\n",
    "\n",
    "### Aligners\n",
    "- STAR\n",
    "- HISAT2\n",
    "- BWA\n",
    "\n",
    "STAR requires a fairly large amount of memory to run compared to all other methods. The index was built using a spacing of 2 (--genomeSAsparseD 2) which technically halves the memory required from around 30GB to 16GB. As a trade-off alignment is slowed down by 25% but results in the identical gene counts. This setting is more cost efficient in most cloud based applications.\n",
    "\n",
    "## Requirements\n",
    "\n",
    "- 26GB of memory\n",
    "- 100GB of free disk space\n",
    "- CPU with high thread count (16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /alignment\n",
    "\n",
    "# create folders\n",
    "mkdir -p sradata\n",
    "mkdir -p quant/combined\n",
    "mkdir -p quant/salmon/\n",
    "mkdir -p quant/kallisto/\n",
    "mkdir -p quant/star/\n",
    "mkdir -p quant/hisat2/\n",
    "mkdir -p time\n",
    "\n",
    "# when using high thread count in STAR there will be a large number of open files. This can lead to errors\n",
    "# set ulimit -n 1024 and it should work with many threads\n",
    "ulimit -n 1024\n",
    "\n",
    "# before we start downloading data we should disable the caching of SRA files\n",
    "# when downloading many files this will quickly end up filling up all disk space otherwise\n",
    "# this command below magically works to disable caching\n",
    "mkdir -p ~/.ncbi\n",
    "echo '/repository/user/cache-disabled = \"true\"' > ~/.ncbi/user-settings.mkfg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the species that you want the SRA samples to be aligned against and the location of precomputed index files. From \"mssm-genecount-combined\" species \"human\" and \"mouse\" can be retrieved. There is a Jupyter notebook to build new index files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SPECIES=\"human\"\n",
    "AWS_BUCKET=\"mssm-genecount-combined\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download pre-computed index files for human. Precomputed mouse index files are also available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scripts/load_index_species.sh ${SPECIES} ${AWS_BUCKET}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 4 helper functions. \"center\" just displays the current state of alignment, \"downloadSRA\" will use faster-dump to download and extract SRA files to fastq. If the sample is based on paired reads it will create two fastq files. \"alignPairedSRA\" and \"alignSingleSRA\" will align the previously downloaded fastq file to the aligners Salmon, kallisto, STAR, HISAT2, and BWA.\n",
    "\n",
    "The resulting quantification counts are then aggregated by an R script into a tsv file for overlapping genes, with columns representing the 5 aligners."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "center() {\n",
    "  echo \"---------------------------------- ${1} ----------------------------------\"\n",
    "}\n",
    "\n",
    "#small helper function for downloads\n",
    "downloadSRA(){\n",
    "    center \"Download SRA file\"\n",
    "    rm -f sradata/*\n",
    "    fasterq-dump \\\n",
    "        --split-files \\\n",
    "        --outdir $2 \\\n",
    "        -e $thread_num -t \"sradata\" \\\n",
    "        $1\n",
    "}\n",
    "\n",
    "alignPairedSRA(){\n",
    "    SRA_FILE=$1\n",
    "    SPECIES=$2\n",
    "    \n",
    "    center \"Salmon\"\n",
    "    \n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo salmon >> time/${SRA_FILE}_time.txt\n",
    "        salmon quant -i \"index/salmon/salmon_${SPECIES}_96\" -l A \\\n",
    "            -1 \"sradata/${SRA_FILE}_1.fastq\" \\\n",
    "            -2 \"sradata/${SRA_FILE}_2.fastq\" \\\n",
    "            -p $thread_num -q --validateMappings -o quant/salmon/$SRA_FILE\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"kallisto\"\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo kallisto >> time/${SRA_FILE}_time.txt\n",
    "        kallisto quant -i index/kallisto/kallisto_${SPECIES}_96.idx \\\n",
    "            \"sradata/${SRA_FILE}_1.fastq\" \\\n",
    "            \"sradata/${SRA_FILE}_2.fastq\" \\\n",
    "            -t $thread_num -o quant/kallisto/$SRA_FILE\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"STAR\"\n",
    "    mkdir -p quant/star/$SRA_FILE\n",
    "    \n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo star >> time/${SRA_FILE}_time.txt\n",
    "        STAR \\\n",
    "            --genomeDir \"index/star/${SPECIES}_96\" \\\n",
    "            --limitBAMsortRAM 10000000000 \\\n",
    "            --runThreadN $thread_num \\\n",
    "            --outSAMstrandField intronMotif \\\n",
    "            --outFilterIntronMotifs RemoveNoncanonical \\\n",
    "            --outFileNamePrefix quant/star/$SRA_FILE/$SRA_FILE \\\n",
    "            --readFilesIn \"sradata/${SRA_FILE}_1.fastq\" \"sradata/${SRA_FILE}_2.fastq\" \\\n",
    "            --outSAMtype BAM SortedByCoordinate \\\n",
    "            --outReadsUnmapped Fastx \\\n",
    "            --outSAMmode Full \\\n",
    "            --quantMode GeneCounts \\\n",
    "            --limitIObufferSize 50000000\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"HISAT2\"\n",
    "    mkdir -p quant/hisat2/$SRA_FILE\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo hisat2 >> time/${SRA_FILE}_time.txt\n",
    "        hisat2 \\\n",
    "            -x \"index/hisat2/${SPECIES}_96/${SPECIES}\" \\\n",
    "            -1 \"sradata/${SRA_FILE}_1.fastq\" \\\n",
    "            -2 \"sradata/${SRA_FILE}_2.fastq\" \\\n",
    "            -p $thread_num \\\n",
    "            -S \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "        featureCounts -T $thread_num -a \"reference/${SPECIES}_96.gtf\" -o \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.tsv\" \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"BWA\"\n",
    "    mkdir -p quant/bwa/$SRA_FILE\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo bwa >> time/${SRA_FILE}_time.txt\n",
    "        bwa mem \\\n",
    "            -t $thread_num \\\n",
    "            \"index/bwa/${SPECIES}_96/${SPECIES}_96\" \\\n",
    "            \"sradata/${SRA_FILE}_1.fastq\" \\\n",
    "            \"sradata/${SRA_FILE}_2.fastq\" \\\n",
    "            > \"quant/bwa/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "        featureCounts -T $thread_num -a \"reference/${SPECIES}_96.gtf\" -o \"quant/bwa/${SRA_FILE}/${SRA_FILE}.tsv\" \"quant/bwa/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    Rscript --vanilla scripts/aggregatecounts.r $SRA_FILE $SPECIES\n",
    "}\n",
    "\n",
    "alignSingleSRA(){\n",
    "    SRA_FILE=$1\n",
    "    SPECIES=$2\n",
    "    \n",
    "    center \"Salmon\"\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo Salmon >> time/${SRA_FILE}_time.txt\n",
    "        salmon quant -i \"index/salmon/salmon_${SPECIES}_96\" -l A \\\n",
    "            -r \"sradata/${SRA_FILE}.fastq\" \\\n",
    "            -p $thread_num -q --validateMappings -o quant/salmon/$SRA_FILE\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"kallisto\"\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo kallisto >> time/${SRA_FILE}_time.txt\n",
    "        kallisto quant -i index/kallisto/kallisto_${SPECIES}_96.idx \\\n",
    "            -t $thread_num -o quant/kallisto/$SRA_FILE \\\n",
    "            --single -l 180 -s 20 \"sradata/${SRA_FILE}.fastq\"\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"STAR\"\n",
    "    mkdir -p quant/star/$SRA_FILE\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo STAR >> time/${SRA_FILE}_time.txt\n",
    "        STAR \\\n",
    "            --genomeDir \"index/star/${SPECIES}_96\" \\\n",
    "            --limitBAMsortRAM 10000000000 \\\n",
    "            --runThreadN $thread_num \\\n",
    "            --outSAMstrandField intronMotif \\\n",
    "            --outFilterIntronMotifs RemoveNoncanonical \\\n",
    "            --outFileNamePrefix quant/star/$SRA_FILE/$SRA_FILE \\\n",
    "            --readFilesIn \"sradata/${SRA_FILE}.fastq\" \\\n",
    "            --outSAMtype BAM SortedByCoordinate \\\n",
    "            --outReadsUnmapped Fastx \\\n",
    "            --outSAMmode Full \\\n",
    "            --quantMode GeneCounts \\\n",
    "            --limitIObufferSize 50000000\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    center \"HISAT2\"\n",
    "    mkdir -p quant/hisat2/$SRA_FILE\n",
    "    { time {\n",
    "        echo \"-----------\" >> time/${SRA_FILE}_time.txt\n",
    "        echo hisat2 >> time/${SRA_FILE}_time.txt\n",
    "        hisat2 \\\n",
    "            -x \"index/hisat2/${SPECIES}_96/${SPECIES}\" \\\n",
    "            -U \"sradata/${SRA_FILE}.fastq\" \\\n",
    "            -p $thread_num \\\n",
    "            -S \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "        featureCounts -T thread_num -a \"reference/${SPECIES}_96.gtf\" -o \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.tsv\" \"quant/hisat2/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    mkdir -p quant/bwa/$SRA_FILE\n",
    "    { time {\n",
    "        bwa mem \\\n",
    "            -t $thread_num \\\n",
    "            \"index/bwa/${SPECIES}_96/${SPECIES}_96\" \\\n",
    "            \"sradata/${SRA_FILE}.fastq\" \\\n",
    "            > \"quant/bwa/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "        featureCounts -T $thread_num -a \"reference/${SPECIES}_96.gtf\" -o \"quant/bwa/${SRA_FILE}/${SRA_FILE}.tsv\" \"quant/bwa/${SRA_FILE}/${SRA_FILE}.sam\"\n",
    "    } 2> temp.txt ; } 2>> time/${SRA_FILE}_time.txt\n",
    "    \n",
    "    Rscript --vanilla scripts/aggregatecounts.r $SRA_FILE $SPECIES\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file we use for benchmarking the speed of the aligners is SRR2972202 from the sample GSM1962466. The origin of the file is human in vivo single Tcell sequences with Illumina HiSeq 2000 for paired-end reads. The file contains 3,161,833 spots, which is on the smaller size. When using other files with significantly higher read it should be kept in mind that some of the algorithms will take fairly long to finish when using low thread counts.\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/sra?term=SRX1461572\n",
    "\n",
    "Reinius, Björn, et al. \"Analysis of allelic expression patterns in clonal somatic cells by single-cell RNA–seq.\" Nature genetics 48.11 (2016): 1430."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRA_FILE=\"SRR2972202\"\n",
    "time downloadSRA $SRA_FILE \"sradata\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The benchmark will caluclate the alignment for multiple settings of thread count. When choosing high thread counts make sure the system running the benchmark supports the number of threads selected.\n",
    "\n",
    "The time to execute each alignment step will be recorded in a text file and saved under the SRA id in the time folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm -f time/${SRA_FILE}_time.txt\n",
    "echo \"---- file ${SRA_FILE} -------\">> time/${SRA_FILE}_time.txt\n",
    "for i in 1 2 3 4 5 6 8 12 16\n",
    "do\n",
    "    thread_num=$i\n",
    "    echo \"---- thread count $i -------\">> time/${SRA_FILE}_time.txt\n",
    "    time alignPairedSRA $SRA_FILE human\n",
    "done\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
